File: Android/scenes/ktv/src/main/java/io/agora/scene/ktv/widget/MusicSettingDialog.java
package io.agora.scene.ktv.widget;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.SeekBar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.GridLayoutManager;

import java.util.ArrayList;
import java.util.List;

import io.agora.scene.base.component.BaseBottomSheetDialogFragment;
import io.agora.scene.base.component.BaseRecyclerViewAdapter;
import io.agora.scene.base.component.BaseViewBindingFragment;
import io.agora.scene.base.component.OnItemClickListener;
import io.agora.scene.base.utils.UiUtil;
import io.agora.scene.ktv.R;
import io.agora.scene.ktv.bean.EffectVoiceBean;
import io.agora.scene.ktv.databinding.KtvDialogMusicSettingBinding;
import io.agora.scene.ktv.databinding.KtvItemEffectvoiceBinding;
import io.agora.scene.ktv.live.fragment.dialog.BeautyVoiceFragment;
import io.agora.scene.ktv.live.fragment.dialog.EarBackFragment;
import io.agora.scene.ktv.live.holder.EffectVoiceHolder;
import io.agora.scene.widget.DividerDecoration;

/**
 * 控制台
 */
public class MusicSettingDialog extends BaseBottomSheetDialogFragment<KtvDialogMusicSettingBinding> implements OnItemClickListener<EffectVoiceBean> {
    public static final String TAG = "MusicSettingDialog";
    private MusicSettingBean mSetting;
    private Boolean isPause = false;
    private BaseRecyclerViewAdapter<KtvItemEffectvoiceBinding, EffectVoiceBean, EffectVoiceHolder> adapter;

    public MusicSettingDialog(MusicSettingBean mSetting, boolean isPause) {
        this.mSetting = mSetting;
        this.isPause = isPause;
    }

    private int getCurrentPitch(int value) {
        switch (value) {
            case 12:
                return 11;
            case 10:
                return 10;
            case 8:
                return 9;
            case 6:
                return 8;
            case 4:
                return 7;
            case 2:
                return 6;
            case -2:
                return 4;
            case -4:
                return 3;
            case -6:
                return 2;
            case -8:
                return 1;
            case -10:
                return 0;
            case -12:
                return -1;
            default:
            case 0:
                return 5;
        }
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ViewCompat.setOnApplyWindowInsetsListener(
                requireDialog().getWindow().getDecorView(), (v, insets) -> {
                    Insets inset = insets.getInsets(WindowInsetsCompat.Type.systemBars());
                    mBinding.getRoot().setPadding(inset.left, 0, inset.right, inset.bottom);
                    return WindowInsetsCompat.CONSUMED;
                });

        // 升降调
        tuningTone(null);
        //mBinding.switchEar.setChecked(this.mSetting.isEar());

        mBinding.sbVol1.setProgress(this.mSetting.getVolMic());
        mBinding.sbVol2.setProgress(this.mSetting.getVolMusic());

        mBinding.changeToneView.setProgress(getCurrentPitch(mSetting.getToneValue()));
        setSoundMode();
        mBinding.btnToneDownDialogSetting.setOnClickListener(v -> tuningTone(false));
        mBinding.btnToneUpDialogSetting.setOnClickListener(v -> tuningTone(true));

        if (isPause) {
            mBinding.sbRemoteVol.setEnabled(false);
            mBinding.btnRemoteVolumeUpDialogSetting.setEnabled(false);
            mBinding.btnRemoteVolumeDownDialogSetting.setEnabled(false);
            mBinding.sbRemoteVol.setProgress(100);
        } else {
            mBinding.sbRemoteVol.setEnabled(true);
            mBinding.btnRemoteVolumeUpDialogSetting.setEnabled(true);
            mBinding.btnRemoteVolumeDownDialogSetting.setEnabled(true);
            mBinding.sbRemoteVol.setProgress(this.mSetting.getRemoteVolume());
        }

        //mBinding.switchEar.setOnCheckedChangeListener((buttonView, isChecked) -> this.mSetting.setEar(isChecked));
        if (this.mSetting.isEar()) {
            mBinding.switchEar.setText("开启");
        } else {
            mBinding.switchEar.setText("关闭");
        }
        mBinding.switchEar.setOnClickListener(this::showEarBackPage);

        mBinding.btVol1Down.setOnClickListener(v -> tuningMicVolume(false));
        mBinding.btVol1Up.setOnClickListener(v -> tuningMicVolume(true));
        mBinding.sbVol1.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                if (seekBar.isPressed()) {
                    mSetting.setVolMic(i);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        mBinding.btVol2Down.setOnClickListener(v -> tuningMusicVolume(false));
        mBinding.btVol2Up.setOnClickListener(v -> tuningMusicVolume(true));
        mBinding.sbVol2.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                if (seekBar.isPressed()) {
                    mSetting.setVolMusic(i);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
        mBinding.btnRemoteVolumeDownDialogSetting.setOnClickListener(v -> {
            int volume = mSetting.getRemoteVolume();
            int newVolume = volume - 1;
            mBinding.sbRemoteVol.setProgress(newVolume);
        });
        mBinding.btnRemoteVolumeUpDialogSetting.setOnClickListener(v -> {
            int volume = mSetting.getRemoteVolume();
            int newVolume = volume + 1;
            mBinding.sbRemoteVol.setProgress(newVolume);
        });
        mBinding.sbRemoteVol.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                if (seekBar.isPressed()) {
                    mSetting.setRemoteVolume(i);
                    mBinding.sbRemoteVol.setProgress(i);
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        List<EffectVoiceBean> list = new ArrayList<>();
        list.add(new EffectVoiceBean(0, R.mipmap.bg_sound_mode_1, "原声"));
        list.add(new EffectVoiceBean(1, R.mipmap.bg_sound_mode_2, "KTV"));
        list.add(new EffectVoiceBean(2, R.mipmap.bg_sound_mode_3, "演唱会"));
        list.add(new EffectVoiceBean(3, R.mipmap.bg_sound_mode_4, "录音棚"));
        list.add(new EffectVoiceBean(4, R.mipmap.bg_sound_mode_1, "留声机"));
        list.add(new EffectVoiceBean(5, R.mipmap.bg_sound_mode_2, "空旷"));
        list.add(new EffectVoiceBean(6, R.mipmap.bg_sound_mode_3, "空灵"));
        list.add(new EffectVoiceBean(7, R.mipmap.bg_sound_mode_4, "流行"));
        list.add(new EffectVoiceBean(8, R.mipmap.bg_sound_mode_1, "R&B"));
        for (EffectVoiceBean item : list) {
            item.setSelect(mSetting.getEffect() == item.getId());
        }

        adapter = new BaseRecyclerViewAdapter<>(list, this, EffectVoiceHolder.class);

        mBinding.rvVoiceEffectList.setAdapter(adapter);
        mBinding.rvVoiceEffectList.addItemDecoration(new DividerDecoration(10, 20, 0));
    }

    private void setSoundMode() {
        int margin = UiUtil.dp2px(10);
        String[] stringArray = getResources().getStringArray(R.array.ktv_audioPreset);
        for (int i = 0; i < stringArray.length; i++) {
            RadioButton radioButton = (RadioButton) getLayoutInflater().inflate(R.layout.btn_sound_mode, null);
            radioButton.setText(stringArray[i]);
            if (i % 4 == 0) {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode4);
            } else if (i % 3 == 0) {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode3);
            } else if (i % 2 == 0) {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode2);
            } else {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode1);
            }
            mBinding.radioGroup.addView(radioButton);
            ((LinearLayout.LayoutParams) radioButton.getLayoutParams()).setMargins(margin, 0, 0, 0);
            if (0 == i) {
                radioButton.setChecked(true);
            } else if (i == stringArray.length - 1) {
                ((LinearLayout.LayoutParams) radioButton.getLayoutParams()).setMargins(margin, 0, margin, 0);
            }
            int finalI = i;
            radioButton.setOnCheckedChangeListener((compoundButton, b) -> {
                if (b) {
                    mSetting.setEffect(finalI);
                }
            });
        }
        ((RadioButton) mBinding.radioGroup.getChildAt(mSetting.getEffect())).setChecked(true);
    }

    private void showEarBackPage(View v) {
        mBinding.getRoot().removeAllViews();
        BaseViewBindingFragment<?> earBackFragment = new EarBackFragment(mSetting);
        FragmentTransaction ft = getChildFragmentManager().beginTransaction();
        ft.add(mBinding.getRoot().getId(), earBackFragment, EarBackFragment.TAG);
        ft.commit();
    }

    /**
     * IMediaPlayer.java
     * /**
     * Sets the pitch of the current media file.
     * pitch Sets the pitch of the local music file by chromatic scale. The default value is 0,
     * which means keeping the original pitch. The value ranges from -12 to 12, and the pitch value
     * between consecutive values is a chromatic value. The greater the absolute value of this
     * parameter, the higher or lower the pitch of the local music file.
     * *
     * - 0: Success.
     * - < 0: Failure.
     * int setAudioMixingPitch(int pitch);
     *
     * @param toneUp true -> +1 | false -> -1 | null -> update value
     */
    private void tuningTone(Boolean toneUp) {
        int newToneValue = this.mSetting.getToneValue();
        if (toneUp != null) {
            if (toneUp) {
                newToneValue += 2;
            } else {
                newToneValue -= 2;
            }
            if (newToneValue > 12)
                newToneValue = 12;

            if (newToneValue < -12)
                newToneValue = -12;

            if (newToneValue != this.mSetting.getToneValue())
                this.mSetting.setToneValue(newToneValue);
            }
        mBinding.changeToneView.setProgress(newToneValue);
    }

    private void tuningMicVolume(Boolean volumeUp) {
        int newVocalVolume = this.mSetting.getVolMic();
        if (volumeUp) {
            newVocalVolume += 1;
        } else {
            newVocalVolume -= 1;
        }
        if (newVocalVolume > 100)
            newVocalVolume = 100;

        if (newVocalVolume < 0)
            newVocalVolume = 0;
        if (newVocalVolume != this.mSetting.getVolMic()) {
            this.mSetting.setVolMic(newVocalVolume);
        }
        mBinding.sbVol1.setProgress(newVocalVolume);
    }

    private void tuningMusicVolume(Boolean volumeUp) {
        int newMusicVolume = this.mSetting.getVolMusic();
        if (volumeUp) {
            newMusicVolume += 1;
        } else {
            newMusicVolume -= 1;
        }
        if (newMusicVolume > 100)
            newMusicVolume = 100;

        if (newMusicVolume < 0)
            newMusicVolume = 0;
        if (newMusicVolume != this.mSetting.getVolMusic()) {
            this.mSetting.setVolMusic(newMusicVolume);
        }
        mBinding.sbVol2.setProgress(newMusicVolume);
    }

    public void onStopPlayer() {
        mBinding.sbRemoteVol.setProgress(100);
    }

    public void onResumePlayer() {
        mBinding.sbRemoteVol.setProgress(this.mSetting.getRemoteVolume());
    }

    @Override
    public void onItemClick(@NonNull EffectVoiceBean data, View view, int position, long viewType) {
        OnItemClickListener.super.onItemClick(data, view, position, viewType);
        Log.e("liu0223", "onItemClick    " + position);

        for (int i = 0; i < adapter.dataList.size(); i++) {
            adapter.dataList.get(i).setSelect(i == position);
            adapter.notifyItemChanged(i);
        }
        mSetting.setEffect(data.getId());
    }

    public interface Callback {
        void onEarChanged(boolean isEar);

        void onMicVolChanged(int vol);

        void onMusicVolChanged(int vol);

        void onEffectChanged(int effect);

        void onBeautifierPresetChanged(int effect);

        void setAudioEffectParameters(int param1, int param2);

        void onToneChanged(int newToneValue);

        void onRemoteVolumeChanged(int volume);

        void onProfessionalModeChanged(boolean enable);

        void onAECLevelChanged(int level);

        void onLowLatencyModeChanged(boolean enable);

        void onEarBackVolumeChanged(int volume);

        void onEarBackModeChanged(int mode);

        void onAINSModeChanged(int mode);
    }
}

File: Android/scenes/cantata/src/main/java/io/agora/scene/cantata/ui/dialog/MusicSettingDialog.kt
package io.agora.scene.cantata.ui.dialog

import android.annotation.SuppressLint
import android.os.Bundle
import android.view.View
import android.widget.CompoundButton
import android.widget.LinearLayout
import android.widget.RadioButton
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isVisible
import io.agora.scene.base.component.BaseBottomSheetDialogFragment
import io.agora.scene.base.component.BaseRecyclerViewAdapter
import io.agora.scene.base.component.BaseRecyclerViewAdapter.BaseViewHolder
import io.agora.scene.base.component.BaseViewBindingFragment
import io.agora.scene.base.component.OnItemClickListener
import io.agora.scene.base.utils.UiUtil
import io.agora.scene.cantata.R
import io.agora.scene.cantata.databinding.CantataDialogMusicSettingBinding
import io.agora.scene.cantata.databinding.CantataItemEffectvoiceBinding
import io.agora.scene.cantata.ui.fragment.EarBackFragment
import io.agora.scene.widget.DividerDecoration

/**
 * 控制台
 */
class MusicSettingDialog constructor(private val mSetting: MusicSettingBean, private val isPause: Boolean) :
    BaseBottomSheetDialogFragment<CantataDialogMusicSettingBinding>() {

    companion object {
        const val TAG = "MusicSettingDialog"
    }

    private var mEffectAdapter:
            BaseRecyclerViewAdapter<CantataItemEffectvoiceBinding, EffectVoiceBean, EffectVoiceHolder>? = null

    private fun getCurrentPitch(value: Int): Int {
        return when (value) {
            12 -> 11
            10 -> 10
            8 -> 9
            6 -> 8
            4 -> 7
            2 -> 6
            -2 -> 4
            -4 -> 3
            -6 -> 2
            -8 -> 1
            -10 -> 0
            -12 -> -1
            0 -> 5
            else -> 5
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        dialog?.window?.let { window ->
            ViewCompat.setOnApplyWindowInsetsListener(window.decorView) { v: View?, insets: WindowInsetsCompat ->
                val inset = insets.getInsets(WindowInsetsCompat.Type.systemBars())
                mBinding.root.apply {
                    setPadding(inset.left, 0, inset.right, inset.bottom)
                }
                WindowInsetsCompat.CONSUMED
            }
        }


        // 升降调
        tuningTone(null)
        mBinding.sbVol1.progress = mSetting.getVolMic()
        mBinding.sbVol2.progress = mSetting.getVolMusic()
        mBinding.changeToneView.progress = getCurrentPitch(mSetting.getToneValue())
        setSoundMode()
        mBinding.btnToneDownDialogSetting.setOnClickListener { v: View? -> tuningTone(false) }
        mBinding.btnToneUpDialogSetting.setOnClickListener { v: View? -> tuningTone(true) }
        if (isPause) {
            mBinding.sbRemoteVol.isEnabled = false
            mBinding.btnRemoteVolumeUpDialogSetting.isEnabled = false
            mBinding.btnRemoteVolumeDownDialogSetting.isEnabled = false
            mBinding.sbRemoteVol.progress = 100
        } else {
            mBinding.sbRemoteVol.isEnabled = true
            mBinding.btnRemoteVolumeUpDialogSetting.isEnabled = true
            mBinding.btnRemoteVolumeDownDialogSetting.isEnabled = true
            mBinding.sbRemoteVol.progress = mSetting.remoteVolume
        }
        mBinding.root.context.apply {
            if (mSetting.isEar()) {
                mBinding.switchEar.text = getString(R.string.cantata_switch_open)
            } else {
                mBinding.switchEar.text = getString(R.string.cantata_switch_close)
            }
        }

        mBinding.switchEar.setOnClickListener { v: View -> showEarBackPage(v) }

        mBinding.btVol1Down.setOnClickListener { v -> tuningMicVolume(false) }
        mBinding.btVol1Up.setOnClickListener { v -> tuningMicVolume(true) }

        mBinding.sbVol1.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                if (seekBar.isPressed) mSetting.setVolMic(i)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })

        mBinding.btVol2Down.setOnClickListener { v -> tuningMusicVolume(false) }
        mBinding.btVol2Up.setOnClickListener { v -> tuningMusicVolume(true) }

        mBinding.sbVol2.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                if (seekBar.isPressed) mSetting.setVolMusic(i)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })

        mBinding.btnRemoteVolumeDownDialogSetting.setOnClickListener { v: View? ->
            val volume = mSetting.remoteVolume
            val newVolume = volume - 1
            mBinding.sbRemoteVol.progress = newVolume
        }
        mBinding.btnRemoteVolumeUpDialogSetting.setOnClickListener { v: View? ->
            val volume = mSetting.remoteVolume
            val newVolume = volume + 1
            mBinding.sbRemoteVol.progress = newVolume
        }

        mBinding.sbRemoteVol.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                if (seekBar.isPressed) {
                    mSetting.remoteVolume = i
                    mBinding.sbRemoteVol.progress = i
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })

        val list: MutableList<EffectVoiceBean> = ArrayList()
        list.add(EffectVoiceBean(0, R.mipmap.bg_sound_mode_1, "原声"))
        list.add(EffectVoiceBean(1, R.mipmap.bg_sound_mode_2, "KTV"))
        list.add(EffectVoiceBean(2, R.mipmap.bg_sound_mode_3, "演唱会"))
        list.add(EffectVoiceBean(3, R.mipmap.bg_sound_mode_4, "录音棚"))
        list.add(EffectVoiceBean(4, R.mipmap.bg_sound_mode_1, "留声机"))
        list.add(EffectVoiceBean(5, R.mipmap.bg_sound_mode_2, "空旷"))
        list.add(EffectVoiceBean(6, R.mipmap.bg_sound_mode_3, "空灵"))
        list.add(EffectVoiceBean(7, R.mipmap.bg_sound_mode_4, "流行"))
        list.add(EffectVoiceBean(8, R.mipmap.bg_sound_mode_1, "R&B"))
        for (item in list) {
            item.setSelect(mSetting.effect == item.id)
        }

        mEffectAdapter = BaseRecyclerViewAdapter(
            list, object : OnItemClickListener<EffectVoiceBean> {

                override fun onItemClick(data: EffectVoiceBean, view: View?, position: Int, viewType: Long) {
                    super.onItemClick(data, view, position, viewType)
                    mEffectAdapter?.let {
                        for (i in it.dataList.indices) {
                            it.dataList[i].setSelect(i == position)
                            it.notifyItemChanged(i)
                        }
                    }

                    mSetting.effect = data.id
                }
            },
            EffectVoiceHolder::class.java
        )

        mBinding.rvVoiceEffectList.adapter = mEffectAdapter
        mBinding.rvVoiceEffectList.addItemDecoration(DividerDecoration(10, 20, 0))

    }

    private fun setSoundMode() {
        val margin = UiUtil.dp2px(10)
        val stringArray = resources.getStringArray(R.array.cantata_audioPreset)
        for (i in stringArray.indices) {
            val radioButton = layoutInflater.inflate(R.layout.cantata_btn_sound_mode, null) as RadioButton
            radioButton.text = stringArray[i]
            if (i % 4 == 0) {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode4)
            } else if (i % 3 == 0) {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode3)
            } else if (i % 2 == 0) {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode2)
            } else {
                radioButton.setBackgroundResource(R.drawable.bg_rbtn_select_sound_mode1)
            }
            mBinding.radioGroup.addView(radioButton)
            (radioButton.layoutParams as LinearLayout.LayoutParams).setMargins(margin, 0, 0, 0)
            if (0 == i) {
                radioButton.isChecked = true
            } else if (i == stringArray.size - 1) {
                (radioButton.layoutParams as LinearLayout.LayoutParams).setMargins(margin, 0, margin, 0)
            }
            radioButton.setOnCheckedChangeListener { compoundButton: CompoundButton?, b: Boolean ->
                if (b) {
                    mSetting.effect = i
                }
            }
        }
        (mBinding.radioGroup.getChildAt(mSetting.effect) as? RadioButton)?.isChecked = true
    }

    private fun showEarBackPage(v: View) {
        mBinding.root.removeAllViews()
        val earBackFragment: BaseViewBindingFragment<*> = EarBackFragment(mSetting)
        val ft = childFragmentManager.beginTransaction()
        ft.add(mBinding.root.id, earBackFragment, EarBackFragment.TAG)
        ft.commit()
    }

    /**
     * IMediaPlayer.java
     * / **
     * Sets the pitch of the current media file.
     * pitch Sets the pitch of the local music file by chromatic scale. The default value is 0,
     * which means keeping the original pitch. The value ranges from -12 to 12, and the pitch value
     * between consecutive values is a chromatic value. The greater the absolute value of this
     * parameter, the higher or lower the pitch of the local music file.
     * *
     * - 0: Success.
     * - < 0: Failure.
     * int setAudioMixingPitch(int pitch);
     *
     * @param toneUp true -> +1 | false -> -1 | null -> update value
     */
    private fun tuningTone(toneUp: Boolean?) {
        var newToneValue = mSetting.getToneValue()
        toneUp?.let { tone ->
            if (tone) {
                newToneValue += 2
            } else {
                newToneValue -= 2
            }
            if (newToneValue > 12) newToneValue = 12
            if (newToneValue < -12) newToneValue = -12
            if (newToneValue != mSetting.getToneValue()) mSetting.setToneValue(newToneValue)
        }
        mBinding.changeToneView.progress = newToneValue
    }

    private fun tuningMicVolume(volumeUp: Boolean) {
        var newVocalVolume = mSetting.getVolMic()
        if (volumeUp) {
            newVocalVolume += 1
        } else {
            newVocalVolume -= 1
        }
        if (newVocalVolume > 100) newVocalVolume = 100
        if (newVocalVolume < 0) newVocalVolume = 0
        if (newVocalVolume != mSetting.getVolMic()) {
            mSetting.setVolMic(newVocalVolume)
        }
        mBinding.sbVol1.progress = newVocalVolume
    }

    private fun tuningMusicVolume(volumeUp: Boolean) {
        var newMusicVolume = mSetting.getVolMusic()
        if (volumeUp) {
            newMusicVolume += 1
        } else {
            newMusicVolume -= 1
        }
        if (newMusicVolume > 100) newMusicVolume = 100
        if (newMusicVolume < 0) newMusicVolume = 0
        if (newMusicVolume != mSetting.getVolMusic()) {
            mSetting.setVolMusic(newMusicVolume)
        }
        mBinding.sbVol2.progress = newMusicVolume
    }


    fun onStopPlayer() {
        mBinding.sbRemoteVol.progress = 100
    }

    fun onResumePlayer() {
        mBinding.sbRemoteVol.progress = mSetting.remoteVolume
    }


}

class EffectVoiceHolder constructor(mBinding: CantataItemEffectvoiceBinding) :
    BaseViewHolder<CantataItemEffectvoiceBinding, EffectVoiceBean>(mBinding) {
    override fun binding(data: EffectVoiceBean?, selectedIndex: Int) {
        data ?: return
        mBinding.ivBg.setImageResource(data.resId)
        mBinding.tvTitle.text = data.title
        mBinding.select.isVisible = data.isSelect
    }
}